test/test_misc_utils.py

```diff
@@ -293,6 +294,107 @@ def test_virtual_app_patch_json():
+
+    assert error_message in m
+    assert offending_url in m
+    assert body_text in m
+    assert str(wrapped_error) in m
+
+    # And the repr is the same.
+    assert repr(e) == str(e)
+
+    # NOTE: Weirdly, I think we'd have had complete code coverage even without this next test, but that illustrates
+    #  why code coverage counts aren't always the right metric. With different data, the same code paths sometimes
+    #  does different things in ways that code coverage tools don't register. -kmp 21-May-2020
+
+    e2 = VirtualAppError(error_message, offending_url, body_json, wrapped_error)
+    m2 = str(e2)
+
+    assert error_message in m2
+    assert offending_url in m2
+    assert str(body_json) in m2             # body_json will be rendered as a Python dict (e.g., with single quotes)
+    assert not json.dumps(body_json) in m2  # So body_json will NOT be rendered as JSON via json.dumps
+    assert str(wrapped_error) in m2
+
+    # And the repr is the same.
+    assert repr(e2) == str(e2)
+
+
+def test_virtual_app_crud_failure():
+
+    simulated_error_message = "simulated error"
+
+    class FakeTestApp(VirtualAppError):
+
+        def __init__(self, app, environ):
+            ignored(app, environ)
+
+        def get(self, url, **kwargs):
+            raise webtest.AppError(simulated_error_message)
+
+        def post_json(self, url, object, **kwargs):
+            raise webtest.AppError(simulated_error_message)
+
+        def patch_json(self, url, fields, **kwargs):
+            raise webtest.AppError(simulated_error_message)
+
+    with mock.patch.object(VirtualApp, "HELPER_CLASS", FakeTestApp):
+
+        app = FakeApp()
+        environ = {'some': 'stuff'}
+
+        vapp = VirtualApp(app, environ)
+
+        some_url = "http://fixture.4dnucleome.org/some/url"
+
+        operations = [
+            lambda: vapp.get(some_url),
+            lambda: vapp.post_json(some_url, {'a': 1, 'b': 2, 'c': 3}),
+            lambda: vapp.patch_json(some_url, {'b': 5})
+        ]
+
+        for operation in operations:
+            try:
+                operation()
+            except Exception as e:
+                assert isinstance(e, VirtualAppError)  # NOTE: not webtest.AppError, which is what was raised
+                assert str(e.raw_exception) == simulated_error_message
+                # assert isinstance(e.raw_exception, webtest.AppError)
+                assert isinstance(e.raw_exception, str)  # <-- TODO: This seems like a bug to me. -kmp 22-May-2020
+
+
+def test_virtual_app_post_failure():
+
+    class FakeTestApp(VirtualAppError):
+
+        def __init__(self, app, environ):
+            ignored(app, environ)
+
+        def get(self, url, **kwargs):
+            raise webtest.AppError("simulated error")
+
+    with mock.patch.object(VirtualApp, "HELPER_CLASS", FakeTestApp):
+
+        app = FakeApp()
+        environ = {'some': 'stuff'}
+
+        vapp = VirtualApp(app, environ)
+
+        with pytest.raises(VirtualAppError):
```

Comment:
I'm not clear on why (syntactically?) having `FakeTestApp` inherit `VirtualAppError` causes the `raise` in `get` to be raised as a `VirtualAppError`. Something is missing in my understanding here.